api/proto/teleport/legacy/types/types.proto

```diff
@@ -1276,25 +1276,24 @@ message ProvisionTokenSpecV2 {
 // ProvisionTokenSpecV2
 message ProvisionTokenSpecV2TPM {
   message Rule {
+    reserved 2, 3;
     // Description is a human-readable description of the rule. It has no
     // bearing on whether or not a TPM is allowed to join, but can be used
     // to associate a rule with a specific host (e.g the asset tag of the server
     // in which the TPM resides).
     // Example: "build-server-100"
     string Description = 1 [(gogoproto.jsontag) = "description,omitempty"];
-    oneof Identifier {
-      // EKPubHash is the SHA256 hash of the EKPub marshaled in PKIX format
-      // and encoded in hexadecimal. This value will also be checked when a TPM
-      // has submitted an EKCert, and the public key in the EKCert will be used
-      // for this check.
-      // Example: d4b45864d9d6fabfc568d74f26c35ababde2105337d7af9a6605e1c56c891aa6
-      string EKPubHash = 2 [(gogoproto.jsontag) = "ekpub_hash,omitempty"];
-      // EKCertSerial is the serial number of the EKCert in hexadecimal with
-      // colon separated nibbles. This value will not be checked when a TPM
-      // does not have an EKCert configured.
-      // Example: 73:df:dc:bd:af:ef:8a:d8:15:2e:96:71:7a:3e:7f:a4
-      string EKCertSerial = 3 [(gogoproto.jsontag) = "ekcert_serial,omitempty"];
-    }
+    // EKPubHash is the SHA256 hash of the EKPub marshaled in PKIX format
+    // and encoded in hexadecimal. This value will also be checked when a TPM
+    // has submitted an EKCert, and the public key in the EKCert will be used
+    // for this check.
+    // Example: d4b45864d9d6fabfc568d74f26c35ababde2105337d7af9a6605e1c56c891aa6
+    string EKPubHash = 4 [(gogoproto.jsontag) = "ekpub_hash,omitempty"];
```

Comment:
This is alright, but what is the problem with oneofs that you are trying to avoid? I don't think oneofs are always a good choice, but I can't say I've had issues with it.

Reply:
Hm - perhaps I'm overly worried. I saw some comments referring to how different JSONpb impls handle oneof's and a quick inspection revealed that basically no other Teleport resources leverage oneof's.

In addition, when I thought this over, I realised that maybe there's a valid intention in specifying both - e.g I want to limit joining to both a specific EKCert serial and an EKPub. I can't see this being particularly useful day to day, but it could be more interesting if we expanded TPM joining to support things like LDevID certs in future which are installed by the Owner rather than the Platform.

Reply:
> I realised that maybe there's a valid intention in specifying both

Yep, that's the real caveat of oneofs, realizing that you do want those fields together sometimes.